sklearn/linear_model/_base.py

```diff
@@ -350,7 +350,7 @@ def predict(self, X):
         xp, _ = get_namespace(X)
         scores = self.decision_function(X)
         if len(scores.shape) == 1:
-            indices = xp.astype(scores > 0, int)
+            indices = xp.astype(scores > 0, xp.int64)
```

Comment:
This change is apparently causing our 32 bit CI to fail with:

```python
TypeError: Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
```

when running on regular numpy inputs.

I can reproduce it on https://scikit-learn.org/stable/lite/lab/ running Pyodide which is a 32 bit Python platform with the following:

```python
import numpy as np
np.take(np.ones(1), np.asarray([0], dtype=np.int64))
```

@rgommers @thomasjpfan  I am not sure what to do here. Should we override our `take` method of our `_NumPyAPIWrapper` wrapper to do convert the indices to the precision matching the platform's bitness?

Reply:
or maybe we could just always use `xp.int32` here? That would restrict scikit-learn classifier to make predictions below 2B data points at a time. I am pretty sure no user has ever reached this.

Reply:
Is there a standard way to inspect the maximum integer index size of an Array API namespace at runtime?

Alternatively we could define something like:

```python
def _max_memory_index_dtype(xp):
    try:
        xp.take(xp.zeros(1), xp.zeros(1, dtype=xp.int64))
        return xp.int64
    except TypeError:
        return xp.int32
```

but that's a big ugly and would require a per-namespace caching mechanism to hide the overhead but it would make it even uglier.

Reply:
Let's instead consider the following suggestion to keep things simpler to maintain:

https://github.com/scikit-learn/scikit-learn/pull/28555/files#r1521513927

Reply:
From [Default Data Types of the Array API](https://data-apis.org/array-api/latest/API_specification/data_types.html#default-data-types):
> A conforming implementation of the array API standard must define the following default data types.
> ...
> - a default array index data type (either int32 or int64).
>
> ...
> The default data types should be clearly defined in a conforming library’s documentation.

This is a bit unfortunate. I would have expected an actual API like `intp`.

Reply:
But there is also this requirement that I overlooked:

> The default array index data type may be int32 on 32-bit platforms, but the default should be int64 otherwise.

So we can use the following cheap idiom: `xp_intp = xp.array([1]).dtype`.

It's a bit magical, but based on my understanding of the spec, it should always be what we are looking for.

Reply:
Those are only "may be" statements, not "must be" statements though.

Reply:
A great and dead simple hack. That might become the new standard to get the default int dtype 🤞

Reply:
(I'll comment here to keep the discussion in one place)

Is the `xp.asarray(1).dtype` hack the way to answer the question "what is the default (and largest) integer data type?". If yes then we could use a hack for now and then start using https://github.com/data-apis/array-api/issues/640?

So maybe for now we can be pragmatic and leave a TODO to revisit it later when the inspection API is widely available?

Reply:
The question would then be: do we add an `def intp_dtype(xp)` helper to factorize and document this hack in a centralized place in our code base?

@rgommers I can do take over this PR to do this and other things if you are busy on other projects. Just let me know.

Reply:
@ogrisel that would be great, thank you 🙏🏼. I can pick it up again if needed, but I'm pretty busy indeed trying to deal with numpy 2.0 and y'all have to make the final decisions here anyway.

Reply:
Implemented in 46a3936.